src/michael_scott_queue.ml

```diff
@@ -55,13 +65,14 @@ let push q v =
       | Nil -> find_tail_and_enq curr_end node
       | Next (_, n) -> find_tail_and_enq n node
   in
-  let newnode = Next (v, Atomic.make Nil) in
-  let tail = Atomic.get q.tail in
-  match tail with
-  | Nil -> failwith "HW_MSQueue.push: impossible"
-  | Next (_, n) ->
-      find_tail_and_enq n newnode;
-      ignore (Atomic.compare_and_set q.tail tail newnode)
+  let new_tail = Atomic.make Nil in
+  let newnode = Next (v, new_tail) in
+  let old_tail = Atomic.get q.tail in
+  find_tail_and_enq old_tail newnode;
+  if Atomic.compare_and_set q.tail old_tail new_tail then
+    match Atomic.get new_tail with
+    | Nil -> ()
+    | Next (_, new_new_tail) -> fix_tail q.tail new_tail new_new_tail
```

Comment:
If we get to this point, it means that we won the race to update the tail.  However, someone else has already queued another element into the queue.  The protocol chosen here is then that the winner must make sure that the tail is in sync, so we call `fix_tail`, which keeps on updating the tail for as long as it keeps on winning.